src/analyses/variable-sensitivity/abstract_pointer_object.cpp

```diff
@@ -74,48 +73,79 @@ abstract_object_pointert abstract_pointer_objectt::write(
+void abstract_pointer_objectt::get_statistics(
+  abstract_object_statisticst &statistics,
+  abstract_object_visitedt &visited,
   const abstract_environmentt &env,
   const namespacet &ns) const
 {
-  pointer_typet pointer_type(to_pointer_type(type()));
-  const typet &pointed_to_type = pointer_type.subtype();
+  abstract_objectt::get_statistics(statistics, visited, env, ns);
+  ++statistics.number_of_pointers;
+}
 
-  return env.abstract_object_factory(pointed_to_type, ns, true, false);
+abstract_object_pointert abstract_pointer_objectt::typecast_from_void_ptr(
+  const exprt &expr,
+  const std::vector<abstract_object_pointert> &operands,
+  const abstract_environmentt &environment,
+  const namespacet &ns) const
+{
+  auto pointer =
+    std::dynamic_pointer_cast<const abstract_pointer_objectt>(operands.front());
+  if(pointer)
+    return pointer->typecast(expr.type(), environment, ns);
+
+  return abstract_objectt::expression_transform(
+    expr, operands, environment, ns);
 }
 
-abstract_object_pointert abstract_pointer_objectt::write_dereference(
-  abstract_environmentt &env,
-  const namespacet &ns,
-  const std::stack<exprt> &stack,
-  const abstract_object_pointert &value,
-  bool merging_write) const
+abstract_object_pointert abstract_pointer_objectt::eval_ptr_diff(
+  const exprt &expr,
+  const std::vector<abstract_object_pointert> &operands,
+  const abstract_environmentt &environment,
+  const namespacet &ns) const
 {
-  if(is_top() || is_bottom())
-  {
-    env.havoc("Writing to a 2value pointer");
-    return shared_from_this();
-  }
+  if(is_top() || operands[1]->is_top())
+    return environment.eval(nil_exprt(), ns);
 
-  return std::make_shared<abstract_pointer_objectt>(type(), true, false);
+  return ptr_diff(expr, operands, environment, ns);
 }
 
-abstract_object_pointert abstract_pointer_objectt::typecast(
-  const typet &new_type,
+abstract_object_pointert abstract_pointer_objectt::eval_ptr_comparison(
+  const exprt &expr,
+  const std::vector<abstract_object_pointert> &operands,
   const abstract_environmentt &environment,
   const namespacet &ns) const
 {
-  INVARIANT(is_void_pointer(type()), "Only allow pointer casting from void*");
-  return std::make_shared<abstract_pointer_objectt>(
-    new_type, is_top(), is_bottom());
+  auto result = eval_ptr_comparison_as_expr(expr, operands, environment, ns);
+  return environment.eval(result, ns);
 }
 
-void abstract_pointer_objectt::get_statistics(
-  abstract_object_statisticst &statistics,
-  abstract_object_visitedt &visited,
-  const abstract_environmentt &env,
+exprt abstract_pointer_objectt::eval_ptr_comparison_as_expr(
+  const exprt &expr,
+  const std::vector<abstract_object_pointert> &operands,
+  const abstract_environmentt &environment,
   const namespacet &ns) const
 {
-  abstract_objectt::get_statistics(statistics, visited, env, ns);
-  ++statistics.number_of_pointers;
+  if(expr.id() == ID_not)
```

Comment:
It feels rather like breaking conditions into atoms should be done elsewhere.

Reply:
Maybe I got carried away here. I can unwind this.